modules/weaver-tests/src/smithy4s/weavertests/WeaverHttpRequestTestCase.scala

```diff
@@ -0,0 +1,250 @@
+  import ce._
+  import org.http4s.implicits._
+  private val baseUri = uri"http://localhost/"
+
+  private def matchRequest(
+      request: Request[IO],
+      testCase: HttpRequestTestCase
+  )(implicit loc: SourceLocation) = {
+    val bodyAssert = testCase.body
+      .map { expectedBody =>
+        request.bodyText.compile.string.map { responseBody =>
+          assert.eql(expectedBody, responseBody)
+        }
+      }
+      .getOrElse(success.pure[IO])
+
+    val headerAssert =
+      testCase.headers
+        .map { expectedHeaders =>
+          expectedHeaders
+            .map { case (name, value) =>
+              val actual = request.headers.get(CIString(name))
+              actual
+                .map { values =>
+                  val actualValue = values.map(_.value).fold
+                  assert.eql(value, actualValue)
+                }
+                .getOrElse(
+                  failure(s"Header $name was not found in the response.")
+                )
+            }
+            .toList
+            .combineAll
+        }
+        .getOrElse(success)
+
+    val expectedUri = baseUri
+      .withPath(
+        Uri.Path.unsafeFromString(testCase.uri)
+      )
+      .withQueryParams(
+        testCase.queryParams.combineAll.map {
+          _.split("=", 2) match {
+            case Array(k, v) =>
+              (
+                k,
+                Uri.decode(
+                  toDecode = v,
+                  charset = StandardCharsets.UTF_8,
+                  plusIsSpace = true
+                )
+              )
+          }
+        }.toMap
+      )
+
+    val uriAssert = assert.eql(expectedUri, request.uri)
+    val methodAssert = assert.eql(
+      testCase.method.toLowerCase(),
+      request.method.name.toLowerCase()
+    )
+
+    List(
+      bodyAssert,
+      headerAssert.pure[IO],
+      uriAssert.pure[IO],
+      methodAssert.pure[IO]
+    ).combineAll
+  }
+
+  private[weavertests] def makeServerTest[I, E, O, SE, SO](
+      endpoint: Endpoint[Op, I, E, O, SE, SO],
+      testCase: HttpRequestTestCase,
+      inputFromDocument: Document.Decoder[I]
+  ): GeneratedTest = {
+    GeneratedTest(
+      name = endpoint.id.toString + "(server|request): " + testCase.id,
+      assertions = {
+        val dummyOutput = DefaultSchemaVisitor(endpoint.output)
+        val input = testCase.params
+          .map { inputFromDocument.decode(_).liftTo[IO] }
+          .getOrElse(IO.pure(().asInstanceOf[I]))
+        (deferred[I], input).tupled
+          .flatMap { case (requestDeferred, input) =>
+            type R[I_, E_, O_, SE_, SO_] = IO[O_]
+
+            val fakeImpl: smithy4s.Monadic[Alg, IO] =
+              service.transform[R](
+                new smithy4s.Interpreter[Op, IO] {
+                  def apply[I_, E_, O_, SE_, SO_](
+                      op: Op[I_, E_, O_, SE_, SO_]
+                  ): IO[O_] = {
+                    val (in, endpointInternal) = service.endpoint(op)
```

Comment:
we could make this easier by adding an unwrap method on the `Endpoint` interface, that would go like : 

``` 
def unwrap(op: Op[_,_,_,_,_]) : Option[I]
```

As for the output, I think this is one of the rare occurrences where it's fine to run the visitor dynamically (to avoid the cast)